### PR TITLE
Vp 635: mandatory fields for save draft or publish

### PR DESCRIPTION
**Pull request by careykwan**
## I do solemnly swear that I have:
- [x] Run `npm test` and all the tests pass.
- [x] Run `npm run lint` and there are no warnings.
- [x] Not decreased the overall test coverage?
- [x] Included the Jira ID and description in the PR title

## Proposed Changes? 🤔
1. Add validation to mandatory fields 
2. 
3. 

## Additional Info.🧐

Any additional info goes here

### CODE REVIEW

**Comment by avowkind** [comment]
Hi Carey - are we waiting for further changes before you put this for review?

**Comment by careykwan** [comment]
> Hi Carey - are we waiting for further changes before you put this for review?

Hey Andrew, yes sorry I was just writing the unit tests, all done now.

File: components/Op/OpDetailForm.js [hunk 1]
```diff
@@ -371,18 +371,31 @@ class OpDetailForm extends Component {
...
-                  {getFieldDecorator('startDate', {})(
+                <Form.Item
+                  label={opStartDate}
+                  name='Start date'
+                >
+                  {getFieldDecorator('startDate', {
+                    rules: [
+                      {
+                        required: this.state.requiredForPublish,
+                        message: 'Start date is required'
```
Comment (avowkind) [inline components/Op/OpDetailForm.js]: Date is not a required field.
activities can be open and take place in the background e.g. negotiable. 
online activities may be done whenever the volunteer wants to do it.

File: components/Op/OpDetailForm.js [hunk 2]
```diff
@@ -397,8 +410,18 @@ class OpDetailForm extends Component {
...
-                  {getFieldDecorator('endDate', {})(
+                <Form.Item
+                  label={opEndDate}
+                  name='End date'
+                >
+                  {getFieldDecorator('endDate', {
+                    rules: [
+                      {
+                        required: this.state.requiredForPublish,
+                        message: 'End date is required'
```
Comment (avowkind) [inline components/Op/OpDetailForm.js]: end date is not required - even if a start date is present the end date can be left blank if the activity is a single day event.

File: components/Op/OpDetailForm.js [hunk 3]
```diff
@@ -414,7 +437,7 @@ class OpDetailForm extends Component {
                   {getFieldDecorator('location', {
                     rules: [
                       {
-                        required: true,
+                        required: this.state.requiredForPublish,
```
Comment (avowkind) [inline components/Op/OpDetailForm.js]: As location is a select box with a default of online. It is not possible for this value to be missing.  Hence the validator is redundant.

**Comment by avowkind** [comment]
Rather than remove the date validation I suggest we leave it in and wrap it in a flag. - defaulted to false.  Then later we can put the flag in the Activity definition and let the activity designer decide whether dates are mandatory or not.   It would make sense for ITF event to ask people to set a date, but not for a request for help.

File: components/Op/OpDetailForm.js (hunk 1 above)
Comment (careykwan) [reply components/Op/OpDetailForm.js]: Dates were listed as a mandatory requirement in the card....??

File: components/Op/OpDetailForm.js (hunk 3 above)
Comment (careykwan) [reply components/Op/OpDetailForm.js]: I can remove this, but the default is 'online', leaving the validation can prompt people to actively change, the more information the user inputs, the better?

**Comment by avowkind** [comment]
Can you keep the validation prompt without failing the validation if they leave it as online?

**Comment by careykwan** [comment]
No you can't, shall I just remove it or default it to false?

**Comment by avowkind** [comment]
For location - remove it. Its easy enough to put back in when there’s an activity template rule to require it.

**Comment by careykwan** [comment]
ok I will remove location and turn off end date, but will be leaving start date as mandatory?

**Comment by avowkind** [comment]
No start date is optional.  

Activities don’t have to have a fixed date.  If I ask someone to volunteer to find me 20 photos of dogs they can do it anytime.

File: components/Op/OpDetailForm.js (hunk 1 above)
Comment (avowkind) [reply components/Op/OpDetailForm.js]: yeah sorry about that - the card was wrong.

**Pull request merged**